datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -224,6 +226,62 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
     }
 }
 
+#[allow(rustdoc::private_intra_doc_links)]
+/// Canonicalize any BinaryExprs that are not in canonical form
+/// <literal> <op> <col> is rewritten to <col> <op> <literal> (remember to switch the operator)
+/// <col> <op> <literal> is canonical
+/// <col1> <op> <col2> is rewritten so that the name of col1 sorts higher than col2 (b > a would be canonicalized to a < b)
+struct Canonicalizer {}
+
+impl Canonicalizer {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+impl TreeNodeRewriter for Canonicalizer {
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+            // Case 1, <col1> <op> <col2>
+            let mut new_expr = BinaryExpr {
+                left: left.clone(),
+                op: op.clone(),
+                right: right.clone(),
+            };
+            match (left.as_ref(), right.as_ref()) {
```

Comment:
You could even push the `op.swap()` into this match, to remove the nested if blocks, e.g.

```rust
match (left.as_ref(), right.as_ref(), op.swap()) {
    (Expr::Column(_), Expr::Column(_), Some(swapped_op)) => {
        todo!(); // swap based on name, directly using the swapped_op
    }
    (Expr::Literal(_), Expr::Column(_), Some(swapped_op)) => {
        todo!(); // swap to have column on left
    }
    _ => todo!(),
}
```

You could then even add the name check of left vs right as an if condition to the branch